core/src/main/java/brooklyn/entity/rebind/RebindIteration.java

```diff
@@ -902,45 +902,53 @@ protected void setCatalogItemId(BrooklynObject item, String catalogItemId) {
-                        BrooklynClassLoadingContext catalogLoader = CatalogUtils.newClassLoadingContext(managementContext, item);
-                        Maybe<Class<?>> catalogClass = catalogLoader.tryLoadClass(jType);
-                        if (catalogClass.isPresent()) {
-                            return (Class<? extends T>) catalogClass.get();
-                        }
+                CatalogItem<?, ?> catalogItem = rebindContext.lookup().lookupCatalogItem(catalogItemId);
+                if (catalogItem == null && BrooklynFeatureEnablement.isEnabled(BrooklynFeatureEnablement.FEATURE_INFER_CATALOG_ITEM_ON_REBIND)) {
+                    // See https://issues.apache.org/jira/browse/BROOKLYN-149
+                    // This is a dangling reference to the catalog item (which will have been logged by lookupCatalogItem).
+                    // Try loading as any version.
+                    if (CatalogUtils.looksLikeVersionedId(catalogItemId)) {
+                        String symbolicName = CatalogUtils.getIdFromVersionedId(catalogItemId);
+                        catalogItem = rebindContext.lookup().lookupCatalogItem(symbolicName);
                     }
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from either classpath of catalog items");
+                }
+                if (catalogItem != null) {
+                    BrooklynClassLoadingContext loader = CatalogUtils.newClassLoadingContext(managementContext, catalogItem);
+                    return loader.loadClass(jType, bType);
                 } else {
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from classpath");
+                    LOG.warn("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId
+                            +" ("+bType.getSimpleName()+"); will try default class loader");
                 }
-
             }
-        }
+            
+            try {
+                return (Class<T>)reflections.loadClass(jType);
+            } catch (Exception e) {
+                Exceptions.propagateIfFatal(e);
+                LOG.warn("Unable to load "+jType+" using reflections; will try standard context");
+            }
 
-        protected BrooklynClassLoadingContext getLoadingContextFromCatalogItemId(String catalogItemId, ClassLoader classLoader, RebindContext rebindContext) {
-            Preconditions.checkNotNull(catalogItemId, "catalogItemId required (should not be null)");
-            CatalogItem<?, ?> catalogItem = rebindContext.lookup().lookupCatalogItem(catalogItemId);
-            if (catalogItem != null) {
-                return CatalogUtils.newClassLoadingContext(managementContext, catalogItem);
+            if (catalogItemId != null) {
+                throw new IllegalStateException("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId+", or load class from classpath");
```

Comment:
Yes, full context of method is that we've tried all available ways of loading (there are earlier return statements in the method), so will now fail. However, if no catalogItemId was specified and if the backards_compatibility feature was enabled then we'll try to infer a catalog item id.

Not particularly nice logic, having multiple return statements, but is possibly the simplest given the unfortunate number of if-else statements to deal with the different scenarios.